### PR TITLE
DT-2616 search users - add name filter

### PR DESCRIPTION
**Pull request by andymarke**
Adds the ability by name; this is a copy of the existing functionality that allows
* Partial matching on either first name or last name
* Partial matching on username
When the filter passed in contains more than one word it is assumed this is 2 words that is first and last name, both combinations are combined in the search.

* Matching account status

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisuserrolesapi/data/filter/UserFilter.kt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+package uk.gov.justice.digital.hmpps.nomisuserrolesapi.data.filter
+
+data class UserFilter(val localAdministratorUsername: String? = null, val name: String? = null)
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisuserrolesapi/data/filter/UserFilter.kt:3]: moved from Specification

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisuserrolesapi/jpa/specification/UserSpecification.kt [hunk 2]
```diff
@@ -16,25 +22,86 @@ class UserSpecification(private val filter: UserFilter) : Specification<UserPers
     criteriaBuilder: CriteriaBuilder
   ): Predicate? {
     val predicates = mutableListOf<Predicate>()
+    fun or(vararg predicates: Predicate) = criteriaBuilder.or(*predicates)
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisuserrolesapi/jpa/specification/UserSpecification.kt:25]: I'm pretty certain we can make the "join"more fluid and extract this out so it is generalised - just takes a bit of time

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisuserrolesapi/jpa/repository/UserPersonDetailRepositoryTest.kt [hunk 3]
```diff
@@ -141,7 +142,13 @@ class UserPersonDetailRepositoryTest {
         .username("jane.bubbles.wwi.bxi")
         .atPrisons(listOf("WWI", "BXI"))
         .build()
-        .transform { user -> user.copy(activeAndInactiveAdministratorOfUserGroups = user.activeAndInactiveAdministratorOfUserGroups.map(makeWWIInactive)) }
+        .transform { user ->
```
Comment (andymarke) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisuserrolesapi/jpa/repository/UserPersonDetailRepositoryTest.kt:145]: just formatting ...

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisuserrolesapi/resource/UserResourceIntTest.kt [hunk 4]
```diff
@@ -111,18 +111,17 @@ class UserResourceIntTest : IntegrationTestBase() {
       @BeforeEach
       internal fun createUsers() {
         with(dataBuilder) {
-          generalUser().username("abella.moulin").firstName("Abella").lastName("Moulin").atPrison("WWI")
-            .buildAndSave()
-          generalUser().username("marco.rossi").atPrisons(listOf("WWI", "BXI")).inactive().buildAndSave()
-          generalUser().username("mark.bowlan").atPrison("BXI").buildAndSave()
+          generalUser().username("abella.moulin").firstName("ABELLA").lastName("MOULIN").atPrison("WWI").buildAndSave()
```
Comment (andymarke) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisuserrolesapi/resource/UserResourceIntTest.kt:114]: upper cased names to match NOMIS behaviour and find relies on them being uppercase

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisuserrolesapi/config/NomisUserRolesApiExceptionHandler.kt [hunk 5]
```diff
@@ -58,6 +59,20 @@ class NomisUserRolesApiExceptionHandler {
       )
   }
 
+  @ExceptionHandler(TypeMismatchException::class)
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisuserrolesapi/config/NomisUserRolesApiExceptionHandler.kt:62]: deal with enum mismatch

**Review by Mjwillis — APPROVED** [review]
LGTM

**Pull request merged**